internal/broker/broker.go

```diff
@@ -2,15 +2,738 @@
+	if err != nil {
+		castErr := &oauth2.RetrieveError{}
+		if errors.As(err, &castErr) && castErr.Response.StatusCode != http.StatusServiceUnavailable {
+			return authCachedInfo{}, fmt.Errorf("could not refresh token: %v", err)
+		}
+
+		st, err := os.Stat(session.cachePath)
+		if err != nil {
+			return authCachedInfo{}, fmt.Errorf("could not get token file info: %v", err)
+		}
+
+		if st.ModTime().Add(b.offlineExpiration).Before(time.Now()) {
+			return authCachedInfo{}, errors.New("token exceeded offline expiration")
+		}
+
+		// This means we are using cached user information, so we don't need to query the provider for them.
+		tok = cachedInfo.Token
+		loadedInfo.UserInfo = cachedInfo.UserInfo
+	}
+	loadedInfo.Token = tok
+
+	loadedInfo.RawIDToken = cachedInfo.RawIDToken
+	refreshedIDToken, ok := tok.Extra("id_token").(string)
+	if ok {
+		loadedInfo.RawIDToken = refreshedIDToken
+	}
+
+	return loadedInfo, nil
+}
+
+func (b *Broker) fetchUserInfo(ctx context.Context, session *sessionInfo, t *authCachedInfo) (userClaims claims, userGroups []group.Info, err error) {
+	defer func() {
+		// Override the error so that we don't leak information. Also, abstract it for the user.
+		// We still log as error for the admin to get access.
+		if err != nil {
+			slog.Error(fmt.Sprintf("Error when fetching user info: %v", err))
+			err = errors.New("could not fetch user info")
+		}
+	}()
+
+	// If we didn't restore user information from the cache, we need to query the provider for it, which means
+	// we need to validate the token.
+	idToken, err := b.auth.provider.Verifier(&b.auth.oidcCfg).Verify(ctx, t.RawIDToken)
+	if err != nil {
+		return claims{}, nil, fmt.Errorf("could not verify token: %v", err)
+	}
+
+	userGroups, err = b.providerInfo.GetGroups(t.Token)
+	if err != nil {
+		return claims{}, nil, fmt.Errorf("could not get user groups: %v", err)
+	}
+
+	if err := idToken.Claims(&userClaims); err != nil {
+		return claims{}, nil, fmt.Errorf("could not get user info: %v", err)
+	}
+
+	if userClaims.Email == "" {
+		return claims{}, nil, errors.New("invalid claims")
+	}
+
+	if !b.skipUsernameValidation && userClaims.Email != session.username {
+		return claims{}, nil, fmt.Errorf("returned user %q does not match the selected one %q", userClaims.Email, session.username)
+	}
+
+	return userClaims, userGroups, nil
+}
+
+type claims struct {
+	Name              string `json:"name"`
+	PreferredUserName string `json:"preferred_username"`
+	Email             string `json:"email"`
+	EmailVerified     bool   `json:"email_verified"`
+	Sub               string `json:"sub"`
+}
+
+func userInfoFromClaims(userClaims claims, groups []group.Info) (string, error) {
+	user := struct {
+		Name   string
+		UUID   string
+		Home   string
+		Shell  string
+		Gecos  string
+		Groups []group.Info
+	}{
+		Name:   userClaims.Email,
+		UUID:   userClaims.Sub,
+		Home:   "/home/" + userClaims.Email,
```

Comment:
I would avoid touching the `/home/` part, but I can see the admin wanting to nest the directories under another one (e.g.: `/home/{provider_addr}/{user_email}`). I only have one concern: don't we open up too much room for failure by doing this? This was one of the reasons we decided to use email as the "user id" in the first place (to have another entity backing us up in the verification).

(also, snaps do not deal very well with user home directories being outside of /home/, so it could become a mess really quickly if the admin doesn't know what he's doing)